code/lib/builder-vite/src/plugins/externals-plugin.ts

```diff
@@ -0,0 +1,96 @@
+import { join } from 'node:path';
+import { init, parse } from 'es-module-lexer';
+import MagicString from 'magic-string';
+import { emptyDirSync, ensureDir, ensureFile, writeFile } from 'fs-extra';
+import { mergeAlias } from 'vite';
+import type { Alias, Plugin } from 'vite';
+import { globals } from '@storybook/preview/globals';
+
+type SingleGlobalName = keyof typeof globals;
+
+/**
+ * This plugin swaps out imports of pre-bundled storybook preview modules for destructures from global
+ * variables that are added in runtime.mjs.
+ *
+ * For instance:
+ *
+ * ```js
+ * import { useMemo as useMemo2, useEffect as useEffect2 } from "@storybook/preview-api";
+ * ```
+ *
+ * becomes
+ *
+ * ```js
+ * const { useMemo: useMemo2, useEffect: useEffect2 } = __STORYBOOK_MODULE_PREVIEW_API__;
+ * ```
+ *
+ * It is based on existing plugins like https://github.com/crcong/vite-plugin-externals
+ * and https://github.com/eight04/rollup-plugin-external-globals, but simplified to meet our meager needs.
+ */
+export async function externalsPlugin() {
+  await init;
+  return {
+    name: 'storybook:externals-plugin',
+    enforce: 'post',
+    // In dev (serve), we set up aliases to files that we write into node_modules/.cache.
+    async config(config, { command }) {
+      if (command !== 'serve') {
+        return undefined;
+      }
+      const newAlias = mergeAlias([], config.resolve?.alias) as Alias[];
+
+      const cachePath = join(process.cwd(), 'node_modules', '.cache', 'vite-plugin-externals');
+      await ensureDir(cachePath);
+      await emptyDirSync(cachePath);
+
+      // eslint-disable-next-line no-restricted-syntax
+      for await (const externalKey of Object.keys(globals) as Array<keyof typeof globals>) {
+        const externalCachePath = join(cachePath, `${externalKey}.js`);
+        newAlias.push({ find: new RegExp(`^${externalKey}$`), replacement: externalCachePath });
+        await ensureFile(externalCachePath);
+        await writeFile(externalCachePath, `module.exports = ${globals[externalKey]};`);
+      }
```

Comment:
This runs each in series, it awaits at every operation.

```
ensure 'a'
write 'a'
ensure 'b'
write 'b'
... etc
```

I think we can do better with a `Promise.all` running all this in parallel.
If this list of files gets larger, performance will suffer.

Key to performance here is to do as much work in parallel as possible.